columnar/src/backend/columnar/columnar_tableam.c

```diff
@@ -2842,4 +2851,406 @@ downgrade_columnar_storage(PG_FUNCTION_ARGS)
+
+		float percentageOfDeleteRows = 
+			(float)stripeDeletedRows / (float)(stripeMetadata->rowCount);
+
+		/* 
+		 * If inspected stripe has less than 0.5 percent of maximum strip row size 
+		 * or percentage of deleted rows is less than 20% we will skip this stripe
+		 * for vacuum.
+		*/
+		if ((stripeMetadata->rowCount > columnarOptions.stripeRowCount * 0.5) &&
+			percentageOfDeleteRows <= 0.2f)
+		{
+			continue;
+		}
+
+		StripeVacuumCandidate *vacuumCandidate = palloc(sizeof(StripeVacuumCandidate));
+		vacuumCandidate->stripeMetadataIndex = stripeMetadataIndex;
+		vacuumCandidate->candidateTotalSize = stripeMetadata->rowCount - stripeDeletedRows; 
+		vacuumCandidate->stripeMetadata = stripeMetadata;
+		vacuumCandidate->activeRows = stripeMetadata->rowCount - stripeDeletedRows;
+
+		vacuumCandidatesStripeList = lappend(vacuumCandidatesStripeList, vacuumCandidate);
+	}
+
+	/* we need all columns */
+	int natts = rel->rd_att->natts;
+	Bitmapset *attr_needed = bms_add_range(NULL, 0, natts - 1);
+
+	/* No quals for table rewrite */
+	List *scanQual = NIL;
+
+	/* Use SnapshotAny when re-writing table as heapAM does */
+	Snapshot snapshot = SnapshotAny;
+
+	MemoryContext scanContext = CreateColumnarScanMemoryContext();
+	bool randomAccess = true;
+
+	ConditionalLockRelationWithTimeout(rel, RowExclusiveLock,
+											VACUUM_TRUNCATE_LOCK_TIMEOUT,
+											VACUUM_TRUNCATE_LOCK_WAIT_INTERVAL,
+											true);
+
+	ColumnarWriteState *writeState = ColumnarBeginWrite(rel->rd_node,
+											columnarOptions,
+											tupleDesc);
+
+	/*
+	 * Combine the vacuum candidates into their own stripes appended to the rel, this
+	 * should clear out any space from partial stripes to make space to move stripes into.
+	 */
+	foreach(lc, vacuumCandidatesStripeList)
+	{
+		StripeVacuumCandidate *vacuumCandidate = lfirst(lc);
+
+		ColumnarReadState *readState = init_columnar_read_state(rel, tupleDesc,
+															attr_needed, scanQual,
+															scanContext, snapshot,
+															randomAccess,
+															NULL);
+
+
+		ColumnarSetStripeReadState(readState,
+								vacuumCandidate->stripeMetadata);
+		
+		Datum *values = palloc0(tupleDesc->natts * sizeof(Datum));
+		bool *nulls = palloc0(tupleDesc->natts * sizeof(bool));
+
+		int32 rowCount = 0;
+		/* we don't need to know rowNumber here */
+		while (rowCount < vacuumCandidate->activeRows && ColumnarReadNextRow(readState, values, nulls, NULL))
+		{
+			ColumnarWriteRow(writeState, values, nulls);
+			rowCount++;
+		}
+
+		DeleteMetadataRowsForStripeId(rel->rd_node, vacuumCandidate->stripeMetadata->id);
+		ColumnarEndRead(readState);
+
+		pfree(values);
+		pfree(nulls);
+	}
+	
+	/*
+	 * We have finished the first route of writes, let's drop our lock until we are
+	 * ready for the next stage: compaction.
+	 */
+	ColumnarEndWrite(writeState);
+	UnlockRelation(rel, RowExclusiveLock);
+
+	/* Lock the relation. */
+	ConditionalLockRelationWithTimeout(rel, RowExclusiveLock,
```

Comment:
RowExclusiveLock will not help here; look at lock conflict table (https://www.postgresql.org/docs/current/explicit-locking.html). 

Relation can be still opened for read/delete/update/insert while you are holding this. So, it can happen that you update stripe and if another process reads same table it will fail with error. 

I think that we should hold AccessExclusiveLock. Acquire that lock when you find space, do physical insert and metadata update and than release it. 

Maybe it is also problematic to insert/update/delete stripe when vacuuming is running - so probably there should be some global atomic boolean flag that will indicate that vacuuming is in progress and will return if DML commands are executing.

Reply:
changed back to `AccessExclusiveLock`.

I don't think that we need a global atomic flag, as we're already holding an `AccessExclusiveLock`.

Reply:
Sure that will work but relation will be blocking for complete duration of vacuum process.

Reply:
So this means will be an equivalent to Full Vacuum and table will not be usable while running it ?

Reply:
we could limit the utility by allowing a maximum number of stripes to be moved before the locks are released and the function returns.

that way it could be run multiple times and have eventual compaction.

Reply:
The ideal goal is to have "incremental" vacuum with intermittent locks. If that can be addressed by having a limited number of actions taken before returning, that could allow this function to be called in a loop from another UDF to have intermittent locks. The only question is how would that loop ever know to exit? I suppose the function could return the number of stripes rewritten, and if that returns 0, the loop could exit.

Reply:
incremental vacuum option added.